book/src/types/rust_types.md

```diff
@@ -0,0 +1,246 @@
+placeholder. Similarly, in that same function, the associated type
+`T::Item` might be represented with a placeholder.
+    
+Like application types, placeholder *types* are only known to be
+equal.
+
+However, we choose not to represent placeholder types as type names
+because they need to be created during type unification and other
+operations, and hence that would require treating `TypeName` less opaquely.
+
+Moreover, when proving negative goals, e.g., `not { Implemented(T:
+Trait) }`, placeholders are treated quite differently from application
+types, since they do not (in fact) represent a known type. When
+solving negative goals, placeholderes are replaced with inference
+variables -- the idea is that this goal is only true if there is *no
+type* `T` that implements `Trait`. Therefore, if we can find no
+answeres for `exists<T> { Implemented(T: Trait) }`, then we know that
+the negation is true. (Note that this means that e.g. `forall<X> { X =
+i32 }` is false but so is `forall<X> { not { X = i32 } }`.)
+
+### Inference variables
+
+The `InferenceVar` variant wraps an `InferenceVar` type.  This
+represents a type whose value is being inferred. The value of an
+inference variables may be "known" or "not known", but that state is
+stored externally, in the inference context (see the section on
+inference below).
+
+When equating, inference variables are treated specially in that they
+become bound (or, if they have already been bound, they are replaced
+with their value).
+
+Inference variables are also integral to canonicalization and
+other types.
+
+### Dyn types
+
+The `Dyn` variant wraps a `DynTy` and represents a `dyn Trait`
+type. In chalk, these are represented as an existential type where we
+store the predicates that are known to be true. So a type like `dyn
+Write` would be represented as, effectively, an `exists<T> { T: Write
+}` type.
+  
+When equating, two `dyn P` and `dyn Q` types are equal if `P = Q` --
+i.e., they have the same bounds. Note that -- for this purpose --
+ordering of bounds is significant, and chalk assumes that a canonical
+ordering is established from somewhere else.
+
+**Question:** We should consider whether logical entailment (e.g., `P
+=> Q` and `Q => P`) would be a better basis. This is more of a Rust
+language design question.
+
+There are "automatic" rules for proving that `dyn P: P` and so forth, but
+that is outside the scope of the chalk-ir crate.
+
+*Question:* in rustc, `dyn Foo` has a lifetime bound, I guess we need that in chalk?
+
+### Function pointer types
+
+The `Fn` variant wraps a `FnTy` struct and represents a `fn()` type
+(in other words, a function pointer). In some ways, fn types are like
+application types, but with one crucial difference: they also contain
+a `forall` binder that for lifetimes whose value is determined when
+the function is called. Consider e.g. a type like `fn(&u32)` or --
+more explicitly -- `for<'a> fn(&'a u32)`.
```

Comment:
`forall` or `for`?

Reply:
Rust syntax is `for`.

Reply:
Mixing the Chalk syntax and Rust syntax can get confusing sometimes